### PR TITLE
[FLINK-6756][DataStream API] Provide Rich AsyncFunction to Scala API …

### PR DESCRIPTION
**Pull request by Clarkkkkk**
## What is the purpose of the change

Provide Rich AsyncFunction to Scala API suite

## Brief change log
  - Proivde RichAsyncFunction for Scala API suite
  - Using a ScalaRichAsyncFunctionWrapper to wrap it into a Java RichAsyncFunction

## Verifying this change

The Java Async runtime context has been tested in RichAsyncFunctionTest. And the access to Async runtime context is now covered in method testRichAsyncFunctionRuntimeContext of AsyncDataStreamITCase.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes

### CODE REVIEW

File: flink-streaming-scala/src/test/scala/org/apache/flink/streaming/api/scala/AsyncDataStreamITCase.scala [hunk 1]
```diff
@@ -135,3 +158,19 @@ class MyAsyncFunction extends AsyncFunction[Int, Int] {
     resultFuture.complete(Seq(input * 3))
   }
 }
+
+class MyRichAsyncFunction extends RichAsyncFunction[Int, String] {
+  override def asyncInvoke(input: Int, resultFuture: ResultFuture[String]): Unit = {
```
[the remark this answers is not in the document]
Comment (Clarkkkkk) [reply flink-streaming-scala/src/test/scala/org/apache/flink/streaming/api/scala/AsyncDataStreamITCase.scala]: I was also thinking about that. But here I want to verify the runtime I got here is a RichAsyncFunctionRuntimeContext, which is a private inner class of RichAsyncFunction. So I cannot use isInstanceOf to verify it. Perhaps I could move the inner class out of RichAsyncFunction and made it accessible by test, so I can verify the RuntimeContext.

[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-streaming-scala/src/test/scala/org/apache/flink/streaming/api/scala/AsyncDataStreamITCase.scala]: Sorry for late response. I wouldn't change the visibility scope for `RichAsyncFunctionIterationRuntimeContext`. I think accessing the `RuntimeContext` is enough in this case (which you do by accessing the `getNumberOfParallelSubtasks`).

Please revert that change, and then I think we will be ready to merge this PR.

**Comment by Clarkkkkk** [comment]
cc @dawidwys

**Comment by dawidwys** [comment]
Just copying my message as top level comment:

> Sorry for late response. I wouldn't change the visibility scope for RichAsyncFunctionIterationRuntimeContext. I think accessing the RuntimeContext is enough in this case (which you do by accessing the getNumberOfParallelSubtasks).
>
>Please revert that change, and then I think we will be ready to merge this PR.

**Comment by Clarkkkkk** [comment]
cc @dawidwys, changes in RichAsyncFunction has been reverted. Thank you for your review.

**Comment by dawidwys** [comment]
Thanks for the contribution @Clarkkkkk. Merging.

**Comment by ealtamir** [comment]
Hi, when could we expect to see this commit in a stable release?

**Comment by connord-stripe** [comment]
(@dawidwys) +1 to this in a stable release, this would have helped me out the past few days, I'm having to write my own wrapper around this in the short term.

**Comment by dawidwys** [comment]
Hi @ealtamir @connord-stripe This PR was already merged into master and is going to be part of 1.8 release, which should be released in the upcoming few days.

**Comment by connord-stripe** [comment]
Neat. Thanks!

**Pull request closed without merging**